File: markdown_gate/classifier.py
from __future__ import annotations

import re

from .config import DOC_TYPES, GateConfig
from .frontmatter import split_frontmatter


HEADING_HINTS = [
    (re.compile(r"^\s*#.+(api|接口|endpoint)", re.I), "public-api-doc"),
    (re.compile(r"^\s*#.+(access|接入|guide|指南)", re.I), "access-guide"),
    (re.compile(r"^\s*#.+(plan|方案|implementation)", re.I), "implementation-plan"),
    (re.compile(r"^\s*#.+(issue|问题|需求)", re.I), "issue-description"),
    (re.compile(r"^\s*#.+(adr|architecture decision record)", re.I), "adr"),
    (re.compile(r"^\s*#.+(runbook|操作手册|应急)", re.I), "runbook"),
]


def classify_document(
    path: str,
    text: str,
    config: GateConfig,
    explicit_type: str | None = None,
) -> tuple[str, dict[str, object], str]:
    metadata, body = split_frontmatter(text)

    if explicit_type:
        return _validate_doc_type(explicit_type), metadata, body

    path_type = config.classify_path(path)
    if path_type:
        return path_type, metadata, body

    fm_type = metadata.get("doc_type") or metadata.get("type")
    if isinstance(fm_type, str) and fm_type:
        return _validate_doc_type(fm_type), metadata, body

    for line in body.splitlines()[:20]:
        for pattern, doc_type in HEADING_HINTS:
            if pattern.search(line):
                return doc_type, metadata, body

    return "unknown", metadata, body


def _validate_doc_type(doc_type: str) -> str:
    if doc_type not in DOC_TYPES:
        known = ", ".join(sorted(DOC_TYPES))
        raise ValueError(f"unknown doc_type {doc_type!r}; expected one of: {known}")
    return doc_type

File: src/classifier.ts
import { classifyPath, type GateConfig } from "./config.js";
import { splitFrontmatter } from "./frontmatter.js";
import type { DocumentType, Metadata } from "./model.js";
import { validateDocType } from "./model.js";

const HEADING_HINTS: Array<[RegExp, DocumentType]> = [
  [/^\s*#.+(api|接口|endpoint)/i, "public-api-doc"],
  [/^\s*#.+(access|接入|guide|指南)/i, "access-guide"],
  [/^\s*#.+(plan|方案|implementation)/i, "implementation-plan"],
  [/^\s*#.+(issue|问题|需求)/i, "issue-description"],
  [/^\s*#.+(adr|architecture decision record)/i, "adr"],
  [/^\s*#.+(runbook|操作手册|应急)/i, "runbook"],
];

export interface ClassifiedDocument {
  docType: DocumentType;
  metadata: Metadata;
  body: string;
}

export function classifyDocument(
  path: string,
  text: string,
  config: GateConfig,
  explicitType?: string,
): ClassifiedDocument {
  const { metadata, body } = splitFrontmatter(text);

  if (explicitType) {
    return { docType: validateDocType(explicitType), metadata, body };
  }

  const pathType = classifyPath(config, path);
  if (pathType) {
    return { docType: pathType, metadata, body };
  }

  const fmType = metadata.doc_type ?? metadata.type;
  if (typeof fmType === "string" && fmType) {
    return { docType: validateDocType(fmType), metadata, body };
  }

  for (const line of body.split(/\r?\n/).slice(0, 20)) {
    for (const [pattern, docType] of HEADING_HINTS) {
      if (pattern.test(line)) {
        return { docType, metadata, body };
      }
    }
  }

  return { docType: "unknown", metadata, body };
}
